time-series/time-series-api/src/main/java/com/powsybl/timeseries/AbstractTimeSeries.java

```diff
@@ -117,36 +117,66 @@ public TimeSeriesMetadata getMetadata() {
 
     protected abstract T createTimeSeries(C chunk);
 
-    private void split(C chunk, List<T> splitList, int newChunkSize) {
+    private void split(C chunkToSplit, List<C> splitChunks, int newChunkSize) {
         if (LOGGER.isTraceEnabled()) {
-            LOGGER.trace("Split chunk [{}, {}]", chunk.getOffset(), chunk.getOffset() + chunk.getLength() - 1);
+            LOGGER.trace("Split chunk [{}, {}]", chunkToSplit.getOffset(), chunkToSplit.getOffset() + chunkToSplit.getLength() - 1);
         }
-        if (chunk.getLength() > newChunkSize) {
+
+        boolean usePreviousChunk = false;
+        C previousChunk = splitChunks.isEmpty() ? null : splitChunks.get(splitChunks.size() - 1);
+        int previousChunkSize = 0;
+        //We can complete the previous chunk if 1) it is uncomplete and 2) the current offset is not a multiple of newChunkSize
+        if (previousChunk != null && previousChunk.getLength() < newChunkSize && chunkToSplit.getOffset() % newChunkSize != 0) {
+            usePreviousChunk = true;
+            previousChunkSize = previousChunk.getLength();
+        }
+        int correctedChunkSize = newChunkSize - previousChunkSize;
+
+        if (usePreviousChunk) {
+            LOGGER.trace("Previous output chunk's size is {} ; {} elements can be added.", previousChunkSize, correctedChunkSize);
+        } else {
+            LOGGER.trace("The previous chunk was complete (or there was no previous chunk). Starting a new one.");
+        }
+
+        if (chunkToSplit.getLength() > correctedChunkSize) {
             // compute lower intersection index with new chunk size
-            int newChunkLowIndex = (int) Math.round(0.5f + (double) chunk.getOffset() / newChunkSize) * newChunkSize;
+            int newChunkLowIndex = (int) Math.round(0.5f + (double) chunkToSplit.getOffset() / correctedChunkSize) * correctedChunkSize;
             if (LOGGER.isTraceEnabled()) {
                 LOGGER.trace("   At index {}", newChunkLowIndex);
             }
-            DataChunk.Split<P, C> split = chunk.splitAt(newChunkLowIndex);
+            DataChunk.Split<P, C> split = chunkToSplit.splitAt(newChunkLowIndex);
             if (LOGGER.isTraceEnabled()) {
                 LOGGER.trace("   Adding chunk [{}, {}]", split.getChunk1().getOffset(), split.getChunk1().getOffset() + split.getChunk1().getLength() - 1);
             }
-            splitList.add(createTimeSeries(split.getChunk1()));
-            split(split.getChunk2(), splitList, newChunkSize);
+
+            if (usePreviousChunk) {
+                C mergedChunk = previousChunk.merge(split.getChunk1());
+                splitChunks.remove(previousChunk);
```

Comment:
This is always the last chunk ? remove by index to avoid linearly scanning the arraylist ?

Reply:
Actually, it might be better to design a mutable API. a.merge(b) would modify a in place. This way, in the future, we may be able to implement merging by storing in a the reference to the array in b, and accessing it directly. This would allow to avoid array copies. What do you think ?

Reply:
Yes, you are right, that would be more performant (less array copies, no creation of chunk, and no modification of the chunk list when usePreviouChunk is true).
I could rename the method extend() to be clear about what the API does. Do you think it is worth keeping the extent method merge, as it is now, for other purposes ?